java/core/src/java/org/apache/orc/InMemoryKeystore.java

```diff
@@ -0,0 +1,403 @@
+            "ORC bad encryption parameter for " + key.getKeyName(), e);
+      }
+
+      final byte[] buffer = new byte[secret.getAlgorithm().keyLength()];
+
+      try {
+        cipher.doFinal(buffer);
+      } catch (final IllegalBlockSizeException e) {
+        throw new IllegalStateException(
+            "ORC bad block size for " + key.getKeyName(), e);
+      } catch (final BadPaddingException e) {
+        throw new IllegalStateException(
+            "ORC bad padding for " + key.getKeyName(), e);
+      }
+
+      return new SecretKeySpec(buffer, secret.getAlgorithm().getAlgorithm());
+    } finally {
+      rwl.writeLock().unlock();
+    }
+
+  }
+
+  /**
+   * Function to add keys to the in-memory KeyStore.
+   *
+   * @param keyName   Name of the key to be added
+   * @param algorithm Algorithm used
+   * @param version   Key Version
+   * @param masterKey Master key
+   */
+  public void addKey(String keyName, EncryptionAlgorithm algorithm, int version,
+      byte[] masterKey) throws IOException {
+
+    if (!keyName.equals(StringUtils.lowerCase(keyName))) {
+      throw new IOException("Key " + keyName + " must be lower case.");
+    }
+
+    rwl.writeLock().lock();
+    try {
+
+      if (meta.containsKey(keyName)) {
+        throw new IOException("Key " + keyName + " already exists.");
+      }
+
+      internalAddKey(keyName, algorithm, version, masterKey);
+
+    } finally {
+      rwl.writeLock().unlock();
+    }
+
+  }
+
+  /**
+   * Function that takes care of adding the key.
+   * NOTE: Here, if the key already exists with same name then this function overwrites it,
+   * this is by design and is used by {@link #rollNewVersion(String, byte[])} function.
+   * Also note, this function is not thread safe, thread safety is the
+   * responsibility of the calling function.
+   *
+   * @param keyName   Name of the key to be added
+   * @param algorithm Algorithm used
+   * @param version   Key Version
+   * @param masterKey Master key
+   * @return The new key
+   */
+  private KeyVersion internalAddKey(String keyName,
+      EncryptionAlgorithm algorithm, int version, byte[] masterKey) {
+
+    /* Test weather platform supports the algorithm */
+    if (!SUPPORTS_AES_256 && (algorithm != EncryptionAlgorithm.AES_128)) {
+      algorithm = EncryptionAlgorithm.AES_128;
+    }
+
+    final byte[] buffer = new byte[algorithm.keyLength()];
+    if (algorithm.keyLength() > masterKey.length) {
+
+      System.arraycopy(masterKey, 0, buffer, 0, masterKey.length);
+      /* fill with zeros */
+      Arrays.fill(buffer, masterKey.length, buffer.length - 1, (byte) 0);
+
+    } else {
+      System.arraycopy(masterKey, 0, buffer, 0, algorithm.keyLength());
+    }
+
+    final KeyVersion key = new KeyVersion(keyName, version, algorithm, buffer);
+
+    keys.put(buildVersionName(keyName, version), key);
+    /* our metadata also contains key material, but it should be fine for testing */
+    meta.put(keyName, key);
```

Comment:
If there isn't already an entry for keyName, you should put the new one. If it is a higher version, it should update it. If it is a lower version, it should not.